### PR TITLE
Add HTTP API for datatypes.

### PR DESCRIPTION
**Pull request by seancribbs**
This rewrites `riak_kv_wm_crdt` and ensures that the functionality it exposes is similar to PB. The JSON format of request/response payloads is described in detail at the top of the module, and implemented in `riak_kv_crdt_json`.

### CODE REVIEW

File: src/riak_kv_wm_crdt.erl [hunk 1]
```diff
@@ -19,148 +19,155 @@
...
+
+service_available(RD, Ctx0) ->
+    Ctx = riak_kv_wm_utils:ensure_bucket_type(RD, Ctx0, #ctx.bucket_type),
+    {ok, Client} = riak_kv_wm_utils:get_riak_client(
+                     local, riak_kv_wm_utils:get_client_id(RD)),
+    {true, RD,
+     Ctx#ctx{client=Client,
+             bucket=path_segment_to_bin(bucket, RD),
+             key=path_segment_to_bin(key, RD),
+             crdt_type=proplists:get_value(crdt, wrq:path_info(RD)),
```
Comment (russelldb) [inline src/riak_kv_wm_crdt.erl]: I thought this came from the bucket_type now (that the path segment was just the string "datatypes")? Should this still be in the code?

File: src/riak_kv_wm_crdt.erl [hunk 2]
```diff
@@ -276,174 +237,284 @@ malformed_boolean_param({Idx, Name, Default}, {Result, RD, Ctx}) ->
...
-normalize_rw_param("quorum") -> quorum;
-normalize_rw_param("all") -> all;
-normalize_rw_param(V) -> list_to_integer(V).
+malformed_timeout_param({Result, RD, Ctx}) ->
+    case wrq:get_qs_value("timeout", undefined, RD) of
+        undefined ->
+            {Result, RD, Ctx};
+        TimeoutStr when is_list(TimeoutStr) ->
+            try
+                Timeout = list_to_integer(TimeoutStr),
```
Comment (russelldb) [inline src/riak_kv_wm_crdt.erl]: Should we also check that it is greater than zero?

File: src/riak_kv_wm_crdt.erl [hunk 3]
```diff
@@ -276,174 +237,284 @@ malformed_boolean_param({Idx, Name, Default}, {Result, RD, Ctx}) ->
...
+        true ->
+            forbidden_check_crdt_type(RD, Ctx);
+        false ->
+            handle_common_error(bucket_type_unknown, RD, Ctx)
+    end.
+
+forbidden_check_crdt_type(RD, Ctx=#ctx{bucket_type = <<"default">>,
+                                       bucket=B0,
+                                       key=K0}) ->
+    B = mochiweb_util:quote_plus(B0),
```
Comment (russelldb) [inline src/riak_kv_wm_crdt.erl:308]: I think some comments  / docs about _why_ would be great here for future me when I maintain this code.

File: src/riak_kv_wm_crdt.erl (hunk 1 above)
Comment (seancribbs) [inline src/riak_kv_wm_crdt.erl]: You're right, I missed this.

File: src/riak_kv_wm_crdt.erl (hunk 2 above)
Comment (seancribbs) [inline src/riak_kv_wm_crdt.erl]: Yes, except doesn't `0` disable the timeout? Either way a guard can be added here.

File: src/riak_kv_wm_crdt.erl (hunk 3 above)
Comment (seancribbs) [inline src/riak_kv_wm_crdt.erl:308]: Right-o.

File: src/riak_kv_wm_crdt.erl [hunk 4]
```diff
@@ -276,174 +236,295 @@ malformed_boolean_param({Idx, Name, Default}, {Result, RD, Ctx}) ->
...
-normalize_rw_param("quorum") -> quorum;
-normalize_rw_param("all") -> all;
-normalize_rw_param(V) -> list_to_integer(V).
+malformed_timeout_param({Result, RD, Ctx}) ->
+    case wrq:get_qs_value("timeout", undefined, RD) of
+        undefined ->
+            {Result, RD, Ctx};
+        TimeoutStr when is_list(TimeoutStr) ->
+            try list_to_integer(TimeoutStr) of
+                Timeout when is_integer(Timeout), Timeout > 0 ->
```
Comment (russelldb) [inline src/riak_kv_wm_crdt.erl:252]: Did you figure out if `> 0` is correct? I wonder if `>= 0` is better, as you say, then `0` can mean `infinity`?

File: src/riak_kv_crdt_json.erl [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    case lists:keyfind(Mod, 2, Mods) of
+        false ->
+            Field;
+        {Type, Mod} ->
+            {Name, Type}
+    end.
+
+-spec field_to_json(map_field()) -> binary().
+field_to_json({Name, Type}) when is_binary(Name), is_atom(Type) ->
+    BinType = atom_to_binary(Type, utf8),
```
Comment (russelldb) [inline src/riak_kv_crdt_json.erl:125]: Efficiency nitpick, but I assume this mapping is static, and could be either added to `?MOD_MAP` or some similar look up. But really, a  nitpick.

**Comment by russelldb** [comment]
Ran eunit tests and tested by hand using riak-erlang-http-client#28.
Looks good, works for me. As mentioned, a riak_test would be nice. We can probably generalise the existing test that uses the PB client. :+1:

File: src/riak_kv_wm_crdt.erl (hunk 4 above)
Comment (seancribbs) [inline src/riak_kv_wm_crdt.erl:252]: I think in the other APIs, yes `0` means `infinity`, but this doesn't coerce it. Fixing.

**Pull request merged**